Documentation/design/renderphase.md

```diff
@@ -0,0 +1,133 @@
+
+### InstallConfig Validation
+
+Render needs to perform 3 kinds of validation:
+
+1. Validate configuration options are known. e.g. networking provider is known.
+
+2. Platform configuration validation. This also includes validating all node pools belong to the specified platform.
+
+3. Ensure `master` node pool exists and has valid configuration.
+
+### TLS asset generation
+
+Generate all the TLS assets required for the cluster and their corresponding Kubernetes secrets. All this needs to be go code rather than terraform templates.
+
+Some of the TLS assets that need to be generated are:
+1. root-ca
+2. kube-ca
+3. aggregator-ca
+4. etcd-ca
+5. apiserver
+6. ingress and more
+
+### Rendering operator manifests
+
+1. Historically, the installer had manifests for all the cluster objects that are required to run the operator. For example, deployment manifest for the operator, service account, roles/rolebindings and operator config etc. So changes to the installer were required to test any change in the operator. This meant that the first indication if the operator had broken install was after the operator changes were merged in its repository and then operator was updated in the installer repository.
+
+2. Also, all the cluster objects required by the operator to functions had to be managed by the installer team.
+
+Therefore, to allow operator teams to iterate and control all aspects of an operator, the installer needs a process to delegate the `InstallConfig` object to each operator in render phase to generate iteself and all other cluster objects necessary for it to function correctly.
+
+#### Prerequisites
+
+1. Render cannot assume container runtime on host.
+2. Render needs to be successfully on (at least) Linux and Darwin systems.
+
+#### Solutions
+
+1. Vendor operators
+
+    One way to achieve rendering is to vendor operator's code into installer and hook operator defined render function in installer's Render phase.
+
+    Pros:
+
+    * No extra work for cross-platform compile for various operators.
+    * Single simple go binary for installer.
+
+    Cons:
+
+    * On operator repository, CI needs to update the toml file and run re-vendor to run install test in their PR.
+
+2. Git submodules
+
+    Installer repo uses git submodules to import operators and then asserts **build and run interfaces** between installer and operator to hook operator into installer's Render phase.
+
+    Pros:
+
+    * Installer repo is smaller.
+    * CI is much easier to work as it doesn't have to deal with go vendoring and only git submodules to replace/update the operator in testing.
+
+    Cons:
+
+    * Build interface between installer and operator repo needs to account for cross-platform builds.
+    * Separate binaries for each operator will be generated, packing them into single binary **might** be challenging in go.
+
+**TODO** Add the preferred solution or other approaches.
+
+### Output
```

Comment:
Not sure that we need to call this out here, but some assets we might want to make clear are sensitive / may not want to be committed (unless you really trust your repo). This could be a separate directory, or could just be default .gitignore entries. These assets still need to be retained, just a git repo might not be the best idea (e.g. root CA key).